fix: Prevent malicious websites from connecting to http / websocket server

tmistele opened the pull request:
Simpler alternative to #1093 that does not break any downstream packages.

However, this does _not_ provide protection on multi-user systems where users don't trust each other. It only prevents websites other than `127.0.0.1` from connecting to our servers. So it's a good first step but doesn't cover all scenarios.

One question:

I noticed there's a second `make_http_server` function that explicitly sets `Access-Control-Allow-Origin: *`. What is that server used for? Obviously, the `*` CORS header looks dangerous.

https://github.com/Myriad-Dreamin/tinymist/blob/8e918cb1322c67160a516c7478d6ecd7a444f014/crates/tinymist/src/task/user_action.rs#L227-L264

tmistele commented:
I just realized this probably won't work like this for vscode because vscode sends something like `vscode-webview://<random-stuff>` as `Origin` header so we can't just check `Origin == 127.0.0.1:port`. But it might be enough to check `Origin` == 127.0.0.1:port || `Origin` starts with "vscode-webview://`". I'll have to think about it (don't have more time to work on it right now, maybe later today).

Myriad-Dreamin commented:
I added it because VS Code rejects fetch data from localhost. It should be good with the only place to use and it should be better to put down all allowed origins one by one instead of a simple wildcard. We should change the code in a separated PR.

https://github.com/Myriad-Dreamin/tinymist/blob/8e918cb1322c67160a516c7478d6ecd7a444f014/tools/editor-tools/src/features/tracing.ts#L127

Myriad-Dreamin commented:
The new PR is much simpler, although it only authenticates the client by the origin. 

Just a small question remaining in my view. Will the check of the origin break people who use http proxy of code servers like gitpod and codeserver? These servers may only forward the http request without changing the origin.

- @tani, the author of the [gitpod.ts](https://github.com/Myriad-Dreamin/tinymist/blob/6813a1081e37e53e64cc1663da8daa033c9e7bc0/editors/vscode/src/gitpod.ts#L15-L32)
- @ because of https://github.com/Myriad-Dreamin/tinymist/issues/625

https://github.com/Myriad-Dreamin/tinymist/blob/6813a1081e37e53e64cc1663da8daa033c9e7bc0/editors/vscode/src/gitpod.ts#L14-L32

tmistele commented:
> I added it because VS Code rejects fetch data from localhost. It should be good with the only place to use and it should be better to put down all allowed origins one by one instead of a simple wildcard. We should change the code in a separated PR.

I've added a commit that I think addresses this, but please have a look @Myriad-Dreamin . This allows only VSCode to connect to the trace/profiling http server for now.

Myriad-Dreamin commented:
> The new PR is much simpler, although it only authenticates the client by the origin.
> 
> Just a small question remaining in my view. Will the check of the origin break people who use http proxy of code servers like gitpod and codeserver? These servers may only forward the http request without changing the origin.
> 
> * @tani, the author of the [gitpod.ts](https://github.com/Myriad-Dreamin/tinymist/blob/6813a1081e37e53e64cc1663da8daa033c9e7bc0/editors/vscode/src/gitpod.ts#L15-L32)
> * @ because of [Websocket connection for preview on code-server #625](https://github.com/Myriad-Dreamin/tinymist/issues/625)
> 
> https://github.com/Myriad-Dreamin/tinymist/blob/6813a1081e37e53e64cc1663da8daa033c9e7bc0/editors/vscode/src/gitpod.ts#L14-L32

This is the final blocking part. I guess we should allow the requests from `url.hostname` but I don't know how to test it:

```ts
 url.hostname = 
   url.port.toString() + 
   "-" + 
   process.env.GITPOD_WORKSPACE_ID + 
   "." + 
   process.env.GITPOD_WORKSPACE_CLUSTER_HOST; 
```

@tani @tmistele

tmistele commented:
> This is the final blocking part. I guess we should allow the requests from `url.hostname` but I don't know how to test it:
> 
> ```ts
>  url.hostname = 
>    url.port.toString() + 
>    "-" + 
>    process.env.GITPOD_WORKSPACE_ID + 
>    "." + 
>    process.env.GITPOD_WORKSPACE_CLUSTER_HOST; 
> ``

Could be. But I don't know how gitpod works, so I'm not sure.

Is this essentially a version of vscode that runs on a gitpod server and renders its output to a browser? If so, maybe it sends a `vscode-webview://...` URL as `Origin` header when connecting to the websocket, just like the "normal" vscode does? So the current version of this PR might already work with gitpod?
